### PR TITLE
[merge 2117 first] enforce lock ordering by checking pointer values instead of mutex name

### PR DESCRIPTION
**Pull request by Greg-Griffith**
edit by sickpig: rebase on top of #2117

### CODE REVIEW

File: src/deadlock-detection/lockorder.h [hunk 1]
```diff
@@ -27,9 +28,13 @@ class CLockOrderTracker
     /// @var std::mutex lot_mutex
     /// mutex that is required to be locked before any method accesses any of this classes data members
     std::mutex lot_mutex;
+    /// @var std::map<void*, std::pair<std::string, boolean> >
+    // map for attempting to track the name of the mutex based on its reference. might not always be accurate, boolean
+    // will denote this
+    / std::map<void *, std::pair<std::string, bool> > mapMutexToName;
```
Comment (sickpig) [inline src/deadlock-detection/lockorder.h]: missing `/` right ?

Comment (Greg-Griffith) [reply src/deadlock-detection/lockorder.h]: extra /. it seems the linter messed up the comments by removing / from previous lines and adding it to this one

File: src/deadlock-detection/lockorder.cpp [hunk 2]
```diff
@@ -10,11 +10,32 @@
...
-void CLockOrderTracker::potential_lock_order_issue_detected(const CLockLocation &thisLock,
-    const CLockLocation &otherLock,
+void CLockOrderTracker::potential_lock_order_issue_detected(LockStackEntry &this_lock,
+    LockStackEntry &other_lock,
     const uint64_t &tid)
 {
+    bool possible_misname = false;
+    if (mapMutexToName.count(this_lock.first) == 0)
+    {
+        throw std::logic_error("critical deadlock order error somewhere");
```
Comment (sickpig) [inline src/deadlock-detection/lockorder.cpp:20]: could we add some context to this message? like the name/value of the lock we are evaluating (eg GetMutexName())?

File: src/deadlock-detection/lockorder.cpp [hunk 3]
```diff
@@ -10,11 +10,32 @@
...
+    }
+    possible_misname = possible_misname || mapMutexToName[this_lock.first].second;
+
+    if (mapMutexToName.count(other_lock.first) == 0)
+    {
+        throw std::logic_error("critical deadlock order error somewhere");
```
Comment (sickpig) [inline src/deadlock-detection/lockorder.cpp:26]: same as above

File: src/deadlock-detection/lockorder.h [hunk 4]
```diff
@@ -27,9 +28,12 @@ class CLockOrderTracker
     /// @var std::mutex lot_mutex
     /// mutex that is required to be locked before any method accesses any of this classes data members
     std::mutex lot_mutex;
+    /// @var std::map<void*, std::pair<std::string, boolean> >
+    /// map for attempting to track the name of the mutex based on its reference. might not always be accurate, boolean will denote this
+    std::map<void*, std::pair<std::string, bool> > mapMutexToName;
     /// @var std::map<std::string, std::set<std::string> > seenLockOrders
-    /// key is lockname, value is vector of locknames that have ever been locked while key was locked
-    std::map<std::string, std::set<std::string> > seenLockOrders;
+    /// key is mutex, value is set of mutexes that have ever been locked while key was locked
```
Comment (sickpig) [inline src/deadlock-detection/lockorder.h]: `is a set of mutex`?

File: src/deadlock-detection/threaddeadlock.cpp [hunk 5]
```diff
@@ -160,31 +160,26 @@ void push_lock(void *c, const CLockLocation &locklocation, LockType locktype, Ow
...
-            }
+            // we have seen the lock we are trying to lock before, check ordering
+            lockdata.ordertracker.CheckForConflict(now, heldLocks, tid);
         }
         else
         {
-            lockdata.ordertracker.AddNewLockInfo(lockname, heldLocks);
-            // track this locks exactly locking order info
+            lockdata.ordertracker.AddNewLockInfo(now, heldLocks);
+            // track this locks exactl locking order info
```
Comment (sickpig) [inline src/deadlock-detection/threaddeadlock.cpp]: exactly

File: src/deadlock-detection/threaddeadlock.cpp [hunk 6]
```diff
@@ -160,31 +160,26 @@ void push_lock(void *c, const CLockLocation &locklocation, LockType locktype, Ow
...
             // get the names of those locks
             for (auto &entry : holdingIter->second)
             {
-                heldLocks.push_back(entry.second);
+                heldLocks.push_back(entry);
             }
-            lockdata.ordertracker.AddNewLockInfo(lockname, heldLocks);
-            // track this locks exactly locking order info
+            lockdata.ordertracker.AddNewLockInfo(now, heldLocks);
+            // track this locks exactl locking order info
```
Comment (sickpig) [inline src/deadlock-detection/threaddeadlock.cpp]: exactly

Comment (Greg-Griffith) [reply src/deadlock-detection/threaddeadlock.cpp]: fixed

File: src/deadlock-detection/threaddeadlock.cpp (hunk 5 above)
Comment (Greg-Griffith) [reply src/deadlock-detection/threaddeadlock.cpp]: fixed

File: src/deadlock-detection/lockorder.h (hunk 4 above)
Comment (Greg-Griffith) [reply src/deadlock-detection/lockorder.h]: fixed

File: src/deadlock-detection/lockorder.cpp (hunk 2 above)
Comment (Greg-Griffith) [reply src/deadlock-detection/lockorder.cpp:20]: This is checking for a scenario that should never happen. If it does happen there is logic flaw somewhere else. The issue wouldnt be related to the mutex being searched for, thats why i kept this error vague

Comment (sickpig) [reply src/deadlock-detection/lockorder.cpp:20]: I see

**Pull request merged**